https://issues.redhat.com/browse/ACM-23758 forwarding alerts mcoa (#8428)
open-cluster-management/rhacm-docs: observability/obs_alert_forward.adoc

[#forward-alert-mcoa]
= Forwarding alerts for multicluster observability add-on

> Configure alert forwarding for the multicluster observability add-on to send alerts to external alert management systems from your managed clusters. By default, multicluster observability add-on is not configured to send alerts from Prometheus on your managed clusters to the hub cluster Alertmanager.

Reviewer: It is specifically for sending alerts to the Alertmanager with provide with the Observability stack on the Hub.

> Configure alert forwarding for the multicluster observability add-on to send alerts to external alert management systems from your managed clusters. By default, multicluster observability add-on is not configured to send alerts from Prometheus on your managed clusters to the hub cluster Alertmanager.

Author: ```suggestion
Configure alert forwarding for the multicluster observability add-on to send alerts to the Alertmanager of the Observability stack from your managed clusters. By default, multicluster observability add-on is not configured to send alerts from Prometheus on your managed clusters to the hub cluster Alertmanager.
```

Create a policy to configure alert forwarding for the multicluster observability add-on.

*Required access:* Cluster administrator

.Prerequisites

- You must have the multicluster observability add-on is enabled. For more details, see xref:../observability/obs_mcoa_intro.adoc#enabling-mcoa[Enabling the multicluster observability add-on].

.Procedure

Complete the following steps to create a policy to configure alert forwarding for the multicluster observability add-on:

> . To configure alert forwarding for platform metrics, create a YAML file named `mcoa-alert-forward-platform`. 

Reviewer: ```suggestion
. To configure alert forwarding for platform metrics, create a YAML file named `mcoa-alert-forward-platform.yaml`. 
```

> .. Add the platform config map details to the `objects-templates-raw` specification. Be sure to add the base domain for your hub cluster in the `$hubBaseDomain` parameter. Your YAML file might resemble the following example:

Reviewer: ```suggestion
.. Your YAML file should look like below. Make sure to add the base domain for your hub cluster in the `$hubBaseDomain` parameter. 
```

Reviewer 2: We cannot use directional language or "look like" -- this is why Mikela wrote it the way that it is. Also we cannot use the word "should". Most of this writing changes to fit the IBM Style and Red Hat Style company guidance.

Reviewer 2: We can play with that line a little bit but typically we say:

See the following example:

Your YAML file might resemble the following example:

Something like that.

Reviewer: I reacted mostly on "might". I would prefer language that more strongly say it _has_ to look like this. Anyway that is no big deal, and feel free to adjust so it matches guidelines. Most important is removing the first sentence there.

Author: ack, will update this soon

Reviewer: Btw @swopebe , I really appreciate the explanation of why it doesn't match the guidelines - that is very helpful so we can learn along the way and try to provide the docs in a way that closer matches the guidelines next time :1st_place_medal:

Reviewer 2: Sure thing! We try to get mutual understanding where there is time.

I agree that _might_ is not always the best way to show an example so we can say:

**See the following example:**

That makes it imperative and takes out room for interpretation.

+
[source,yaml]
----
apiVersion: policy.open-cluster-management.io/v1
kind: Policy
metadata:
  name: mcoa-alert-forward-platform
  namespace: open-cluster-management-global-set
spec:
  disabled: false
  policy-templates:
  - objectDefinition:
      apiVersion: policy.open-cluster-management.io/v1
      kind: ConfigurationPolicy
      metadata:
        name: mcoa-alert-forward-platform
      spec:
        namespaceSelector:
          exclude:
          - kube-*
          include:
          - default
        object-templates-raw: |
          {{ $hubBaseDomain:= "INPUT_BASE_DOMAIN" }}
          {{ $hubName := (split "." $hubBaseDomain)._0 }}
          {{- $cmo := (lookup "v1" "ConfigMap" "openshift-monitoring" "cluster-monitoring-config") }}
          {{- $cy := dict }}
          {{- if and $cmo $cmo.data }}
            {{- $cy = (index $cmo "data" "config.yaml") | fromYaml }}
          {{- end }}
----

> .. Define what you want your configuration to include. Your YAML file might resemble the following partial example:

Reviewer: I don't expect people really changes this configuration much. Only thing they have to change is the "INPUT_BASE_DOMAIN" above.

Reviewer 2: I would just merge this with the YAML above. We only need to pull out important specs or other YAML values (if there are many) that need to be changed or built a certain way for part of the procedure. We don't typically pull this info out .

Author: ack, updating

+
[source,yaml]
----
          {{- $mangedConfig := dict }}
 
          {{- $pm := `
            prometheusK8s:
              additionalAlertmanagerConfigs:
              - apiVersion: v2
                bearerToken:
                  key: token
                  name: observability-alertmanager-accessor-%[1]s
                scheme: https
                staticConfigs:
                - alertmanager-open-cluster-management-observability.apps.%[2]s
                tlsConfig:
                  ca:
                    key: service-ca.crt
                    name: hub-alertmanager-router-ca-%[1]s
                  insecureSkipVerify: false
              externalLabels:
                managed_cluster: %[3]s
          ` }}
----

> .. Merge all of the configurations into one. Your YAML file might resemble the following partial example:

Reviewer: I think we should avoid splitting the yaml configuration as that just adds confusion in my opinion. If you need it to be smaller the placementBinding can be separate as you already did.

If it's not possible to keep as one, is there no other way to fix it? Can we provide the file to download instead of rendering it directly on the docs site?

Reviewer: Other places we have collapsible examples like here, would that not be acceptable? https://docs.redhat.com/en/documentation/red_hat_advanced_cluster_management_for_kubernetes/2.4/html-single/governance/index#policy-sample-file

Reviewer 2: @jacobbaungard we are no longer (for a while now) dumping large YAML samples in the doc without explanation or procedures, that is why she did this for this PR, but in this case, these YAML breakouts are not necessary.

2.4 doc is very old, not good to use bc we were not doing peer reviews and we were pointing to links we should not have.

I will go over this with the team again: When to break YAML up for a procedure, when not to break it up. In this case, the config info here does not need to be separated, where if the user is adding specs to YAML, it likely does.

We are no longer allowed to use callouts and with the move to DITA, you can expect we will need to build procedures that way.

Reviewer: Right, copied the wrong link, it is also there in [2.14](https://docs.redhat.com/en/documentation/red_hat_advanced_cluster_management_for_kubernetes/2.14/html-single/governance/index#policy-sample-file).

Author: Thanks Brandi. Yup, this is just my attempt to break up the long YAML a bit.

+
[source,yaml]
----
            {{- $mangedConfig = merge $mangedConfig
                              ((printf $pm $hubName $hubBaseDomain (fromClusterClaim "id.openshift.io"))| fromYaml)
          }}
          - complianceType: mustonlyhave
            objectDefinition:
              apiVersion: v1
              data:
                config.yaml: |
                  {{ (merge $cy $mangedConfig)| toYaml |autoindent }}
              kind: ConfigMap
              metadata:
                name: cluster-monitoring-config
                namespace: openshift-monitoring
            recordDiff: InStatus

>   remediationAction: inforce

Reviewer: ```suggestion
  remediationAction: inform
```

----

> //you mention `inform` mode, but I only see `enforce` for the remediation action. Please clarify. Do we need to mention anything about `inform` mode here?

Reviewer 2: comment is added here, just in case.

-----

For these YAML samples, I could be wrong, but I think we have to add something to the YAML to show it is a part of a larger. See the following example where we add `...`

```
Complete the following steps to add a worker node to the managed cluster:

. Define a new node object in the existing `ClusterInstance` custom resource:

+
[source,yaml]
----
spec:
  ...
  nodes:
    - hostName: "<host_name>"
      role: "worker"
      templateRefs:
        - name: ai-node-templates-v1
          namespace: rhacm
      bmcAddress: "<bmc_address>"
      bmcCredentialsName:
        name: "<bmc_credentials_name>"
      bootMACAddress: "<boot_mac_address>"
...
----
```

> //you mention `inform` mode, but I only see `enforce` for the remediation action. Please clarify. Do we need to mention anything about `inform` mode here?

Reviewer: Sorry, I meant to put them in "inform" in the example. I added suggestions to fix it.

.. Create a `PlacementBinding` resource to bind the policy to the `global` placement. Your YAML file might resemble the following sample:

+
[source,yaml]
----
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
  name: mcoa-alert-forward-platform-placement
  namespace: open-cluster-management-global-set
placementRef:
  apiGroup: cluster.open-cluster-management.io
  kind: Placement
  name: global
subjects:
- apiGroup: policy.open-cluster-management.io
  kind: Policy
  name: mcoa-alert-forward-platform
----

.. Apply your YAML file. If you make custom changes to the `additionalAlertmanagerConfigs` specification of the config map, be sure to add the changes directly into the policy, before applying the policy. Run the following command:

+
[source,bash]
----
oc apply -f ./forward-platform-alerts.yaml
----

+

> *Note:* If you have a managed cluster other than a {ocp} cluster, you must create a `Placement` resource.

Reviewer: ```suggestion
*Note:* If you have managed clusters other than a {ocp} cluster, you must create the appropriate `PlacementBinding` resource targeting the managed clusters.
```

Author: Heads up, i updated your suggestion

> .. To apply the policy changes across all managed clusters and enable alert forwarding from your managed clusters to the hub cluster, change the `remediationAction` parameter value to `enforce`. 

Reviewer: ```suggestion
.. The policy presented here is in inform mode. This means you can validate the changes to be applied using the governance user interface in the OCP console, to ensure you are satisfied with the change. The policy will show up to be as non-complient, and you can then verify the diff the policy would apply if enforced. To apply the policy changes across all managed clusters and enable alert forwarding from your managed clusters to the hub cluster, change the `remediationAction` parameter value to `enforce`. 
```

Anything we can do to highlight this, so it is not missed, would be good.

Reviewer 2: We also cannot say "presented here" -- as it is self-demonstrative. We try to avoid "here". Same with "this means"

Reviewer: sure, just adjust it to correct wording. Most important is the information in the additional couple of sentences are conveyed as that info was dropped from the original text.

Author: ack

Reviewer 2: ok, we have a little work to do here to change without changing the meaning, but I think we can do it pretty easily.

+
[source,yaml]
----
apiVersion: policy.open-cluster-management.io/v1
kind: Policy
metadata:
  name: mcoa-alert-forward-uwl
  namespace: open-cluster-management-global-set
spec:
  disabled: false
  policy-templates:
  - objectDefinition:
      apiVersion: policy.open-cluster-management.io/v1
      kind: ConfigurationPolicy
      metadata:
        name: mcoa-alert-forward-uwl
      spec:
        namespaceSelector:
          exclude:
          - kube-*
          include:
          - default
        object-templates-raw: |
          {{ $hubBaseDomain:= "INPUT_BASE_DOMAIN" }}
          {{ $hubName := (split "." $hubBaseDomain)._0 }}
 
          ## Get the current ConfigMap and read the value as a yaml structure
          {{- $cmo := (lookup "v1" "ConfigMap" "openshift-user-workload-monitoring" "user-workload-monitoring-config") }}
          {{- $cy := dict }}
          {{- if and $cmo $cmo.data }}
            {{- $cy = (index $cmo "data" "config.yaml") | fromYaml }}
          {{- end }}
----

> .. Define what you want your configuration to include. Your YAML file might resemble the following partial example:

Reviewer 2: What is the reason for breaking this out in to a step? Not sure what the goal is, that is why I am asking.

Reviewer 2: Same feedback, we don't need to break this out.

Author: similar to my response. This was my attempt to break up the YAML. The hidden comment from the original YAML mentioned  "Defining what you want your configuration to include.."

                scheme: https
                staticConfigs:
                - alertmanager-open-cluster-management-observability.apps.%[2]s
                tlsConfig:
                  ca:
                    key: service-ca.crt
                    name: hub-alertmanager-router-ca-%[1]s
                  insecureSkipVerify: false
              externalLabels:
                managed_cluster: %[3]s
          ` }}
----

.. Merge all of the configurations into one. Your YAML file might resemble the following partial example:

+
[source,yaml]
----
            {{- $mangedConfig = merge $mangedConfig
                              ((printf $pm $hubName $hubBaseDomain (fromClusterClaim "id.openshift.io"))| fromYaml)
          }}
          - complianceType: mustonlyhave
            objectDefinition:
              apiVersion: v1
              data:
                config.yaml: |
                  {{ (merge $cy $mangedConfig )| toYaml |autoindent }}
              kind: ConfigMap
              metadata:
                name: user-workload-monitoring-config
                namespace: openshift-user-workload-monitoring
            recordDiff: InStatus

>   remediationAction: enforce

Reviewer: ```suggestion
  remediationAction: inform
```
